### PR TITLE
net: l2: {ethernet,ppp}: fix k_work API usage in carrier on/off handling

### PR DESCRIPTION
**Pull request by mniestroj**
`net_{eth,ppp}_carrier_on()` and `net_{eth,ppp}_carrier_off()` 
call `k_work_init()` on work item that can be pending or still 
be processed in another thread. This results in undefined behavior.

Initialize work item once and use an atomic flag to switch between
up/down carrier state. Submit work to workqueue whenever up/down carrier
state changes, so that last state is always properly propagated to
network interface layer.

This is an alternative to #32759.

### CODE REVIEW

**Review by jukkar — COMMENTED** [review]
Looks good, just a minor comment

File: include/net/ethernet.h [hunk 1]
```diff
@@ -415,6 +421,9 @@ struct ethernet_context {
 	int8_t vlan_enabled;
 #endif
 
+	/** Is network carrier up */
+	bool is_net_carrier_up;
```
[the remark this answers is not in the document]
Comment (mniestroj) [reply include/net/ethernet.h:425]:
```
bool is_net_carrier_up : 1;
bool is_init : 1;
```
or 
```
uint8_t is_net_carrier_up : 1;
uint8_t is_init : 1;
```
?

[the remark this answers is not in the document]
Comment (mniestroj) [reply include/net/ethernet.h]: I used `bool`, as it is pretty common in Zephyr tree.

**Review by jukkar — APPROVED** [review]

**Comment by jukkar** [comment]
@jimparis would this fix the issues you were seeing in #32759

**Comment by jimparis** [comment]
Yes, it should.

**Comment by jukkar** [comment]
> Yes, it should.

Great, if you are ok with it, +1 would be appreciated so we get this merged :smile:

**Review by jimparis — APPROVED** [review]

**Pull request merged**